compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java

```diff
@@ -1120,6 +1120,10 @@ private void handleDeclaredWithVar(BLangVariable variable) {
             variable.type = symTable.semanticError;
             dlog.error(variable.pos, DiagnosticErrorCode.VARIABLE_DECL_WITH_VAR_WITHOUT_INITIALIZER);
         } else {
+            if (varRefExpr.getKind() == NodeKind.BINARY_EXPR &&
```

Comment:
This is not an issue only when used with `var`. An additive expression `a + b` should fail if `a` and `b` are of different basic types.

Additive expression
> It is required that either
>
> - the static type of both operand expressions is a subtype of the same basic type, in which case this basic type will be the static type of the result, or
> - the static type of one operand expression must be a subtype of xml and of the other operand expression must be a subtype of string, in which case the static type of the result is xml

Multiplicative expression
> The static type of both operand expressions is required to be a subtype of the same basic type; this basic type will be the static type of the result.

```ballerina
const A = 10;
const B = 20;

type C A|B;

function testImplicitConversion() {
    float a = 1;
    decimal b = 1;
    any i1 = a + b; // No error, should fail
    var i2 = a + b; // ERROR: implicit type conversion not allowed here

    int d = 10;
    float e = 12.25;
    any i3 = d - e; // No error, should fail
    var i4 = d - e; // ERROR: implicit type conversion not allowed here
}
```

Also rather than saying "implicit type conversion not allowed here" I think we should stick to errors similar to the ones we have already "operator '+' not defined for 'float' and 'decimal'". IMO, this will be more helpful to the user.